devtools/x/src/lint/guppy.rs

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! Project and package linters that run queries on guppy.
+
+use crate::config::EnforcedAttributesConfig;
+use std::collections::HashMap;
+use x_lint::prelude::*;
+
+/// Ban certain crates from being used as direct dependencies.
+#[derive(Debug)]
+pub struct BannedDirectDeps<'cfg> {
+    banned_deps: &'cfg HashMap<String, String>,
+}
+
+impl<'cfg> BannedDirectDeps<'cfg> {
+    pub fn new(banned_deps: &'cfg HashMap<String, String>) -> Self {
+        Self { banned_deps }
+    }
+}
+
+impl<'cfg> Linter for BannedDirectDeps<'cfg> {
+    fn name(&self) -> &'static str {
+        "banned-direct-deps"
+    }
+}
+
+// This could be done either as a project linter or as a package linter -- doing it as a package
+// linter is slightly cheaper empirically.
+impl<'cfg> ProjectLinter for BannedDirectDeps<'cfg> {
```

Comment:
You say that doing this as a package linter is cheaper than proceed to impl a ProjectLinter, is the comment inaccurate?

Reply:
Fixed the comment.